### ISSUE TITLE
Failed to send heartbeat error

### ISSUE THREAD

**Issue by steve-chavez** [sozu-proxy/lapin#54]
I'm trying to combine tokio-postgres with lapin-futures(version 0.8.2, using RMQ 3.6.9), the use case is listening to postgres notifications and publishing to a RMQ queue with lapin, this example works:

```rust
fn main() {
  env_logger::init().unwrap();
  let mut core = Core::new().unwrap();
  let handle = core.handle();

  let addr = "127.0.0.1:5672".parse().unwrap();

  core.run(
    TcpStream::connect(&addr, &handle)
      .and_then(|stream| lapin::client::Client::connect(stream, &ConnectionOptions::default()))
      .and_then(|client| client.create_channel())
      .and_then(|channel| {
        channel.queue_declare("hello", &QueueDeclareOptions::default(), FieldTable::new()).and_then(|_| {
          Connection::connect("postgres://postgres@localhost", TlsMode::None, &handle)
          .then(|conn| conn.unwrap().batch_execute("LISTEN pgchannel1").map_err(|_|Error::new(ErrorKind::Other, "")))
          .and_then(|conn|
            conn.notifications().map_err(|_|Error::new(ErrorKind::Other, "")).for_each(move |n|{
              println!("Publishing payload: {}", n.payload);
              channel.basic_publish(
                  "hello",
                  n.payload.as_bytes(),
                  &BasicPublishOptions::default(),
                  BasicProperties::default()
              );
              Ok(())
            })
          )
        })
      })
  ).unwrap();
}
```

But after 3 mins of being idle, the RMQ tcp connection is dropped(disappears from RMQ management web interface), and when trying to send another notification to the queue the following message is shown:

```rust
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 104, message: "Connection reset by peer" } }
ERROR:lapin_futures::transport: Failed to send heartbeat: Error { repr: Os { code: 32, message: "Broken pipe" } }
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
ERROR:lapin_futures::transport: upstream poll gave Ready(None)
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
ERROR:lapin_futures::transport: upstream poll gave Ready(None)
```

I've checked and the 3 mins are because of 3 failed heartbeats(60s each one, same error can be obtained faster by reducing heartbeat). 

How can I keep the lapin connection alive while consuming another stream? Any pointers would help. Thanks.

**Comment by Geal** [comment]
hey, why did you close the issue? If the bug is still present, we'll fix it (it's just that I might not be able to work on it right away)

**Comment by steve-chavez** [comment]
My bad, here is an example without the postgres dependency, just publishing stdin inputs:

```rust
extern crate tokio_core;
extern crate amq_protocol;
extern crate futures;
extern crate env_logger;
extern crate lapin_futures as lapin;

use amq_protocol::types::FieldTable;
use futures::{Future, Sink, Stream};
use futures::stream::BoxStream;
use futures::sync::mpsc::channel;
use lapin::client::{Client, ConnectionOptions};
use lapin::channel::{BasicPublishOptions,QueueDeclareOptions, BasicProperties};
use std::io::{self, BufRead};
use std::thread;
use tokio_core::reactor::Core;
use tokio_core::net::TcpStream;

fn stdin() -> BoxStream<String, io::Error> {
    let (mut tx, rx) = channel(1);
    thread::spawn(move || {
        let input = io::stdin();
        for line in input.lock().lines() {
            match tx.send(line).wait() {
                Ok(s) => tx = s,
                Err(_) => break,
            }
        }
    });
    rx.then(|e| e.unwrap()).boxed()
}

fn main() {
  env_logger::init().unwrap();
  let mut core = Core::new().unwrap();
  let handle = core.handle();
  let addr = "127.0.0.1:5672".parse().unwrap();

  core.run(
    TcpStream::connect(&addr, &handle).and_then(|stream| Client::connect(stream, &ConnectionOptions::default()))
    .and_then(|client| client.create_channel())
    .and_then(|channel|
      channel.queue_declare("test", &QueueDeclareOptions::default(), FieldTable::new())
      .and_then(move |_|
        stdin().for_each(move |string| {
          println!("Sending: {}", string);
          channel.basic_publish("test", string.as_bytes(), &BasicPublishOptions::default(), BasicProperties::default());
          Ok(())
        })
      )
    )
  ).unwrap();
}
```

It gives the same error after 3 mins, I think the same as the previous example is happening: there's blocking in the ```for_each```, maybe an improvement for lapin could be having the heartbeat logic in a background thread.

**Comment by Keruspe** [comment]
Could you paste the output with RUST_LOG=trace ?

**Comment by steve-chavez** [comment]
Sure, this is the output:

```rust
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 28, nanos: 34340949 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 3016, tv_nsec: 485940297 }
TRACE:tokio_core::reactor: event Ready {Readable} Token(1)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 41038 }
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Method(1, Basic(Publish(Publish { ticket: 0, exchange: "", routing_key: "test", mandatory: false, immediate: false })))
TRACE:lapin_futures::transport: serialized frame: 21 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=21
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 104, message: "Connection reset by peer" } }
TRACE:lapin_futures::transport: stream poll
TRACE:lapin_futures::transport: Sending heartbeat
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Heartbeat(0)
TRACE:lapin_futures::transport: serialized frame: 8 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=8
ERROR:lapin_futures::transport: Failed to send heartbeat: Error { repr: Os { code: 32, message: "Broken pipe" } }
TRACE:tokio_io::framed_read: attempting to decode a frame
TRACE:lapin_futures::transport: decoded frame: Heartbeat(0)
TRACE:tokio_io::framed_read: frame decoded from buffer
TRACE:lapin_futures::transport: upstream poll gave frame: Heartbeat(0)
TRACE:lapin_async::connection: will handle frame: Heartbeat(0)
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Heartbeat(0)
TRACE:lapin_futures::transport: serialized frame: 8 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=8
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
TRACE:lapin_futures::transport: stream poll
TRACE:tokio_io::framed_read: attempting to decode a frame
TRACE:lapin_futures::transport: decoded frame: Heartbeat(0)
TRACE:tokio_io::framed_read: frame decoded from buffer
TRACE:lapin_futures::transport: upstream poll gave frame: Heartbeat(0)
TRACE:lapin_async::connection: will handle frame: Heartbeat(0)
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Heartbeat(0)
TRACE:lapin_futures::transport: serialized frame: 8 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=8
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
TRACE:lapin_futures::transport: stream poll
TRACE:tokio_io::framed_read: attempting to decode a frame
ERROR:lapin_futures::transport: upstream poll gave Ready(None)
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Header(1, 60, ContentHeader { class_id: 60, weight: 0, body_size: 9, properties: Properties { content_type: None, content_encoding: None, headers: None, delivery_mode: None, priority: None, correlation_id: None, reply_to: None, expiration: None, message_id: None, timestamp: None, type_: None, user_id: None, app_id: None, cluster_id: None } })
TRACE:lapin_futures::transport: serialized frame: 22 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=22
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
TRACE:lapin_futures::transport: stream poll
ERROR:lapin_futures::transport: upstream poll gave Ready(None)
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 6532 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 3016, tv_nsec: 496110091 }
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 7443 }
```

**Comment by Geal** [comment]
@steve-chavez in the logs, there's a `ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 104, message: "Connection reset by peer" } }` so the error happens much sooner. Now I need to check why we don't stop when we get this error, but it's not coming from the heartbeat

**Comment by Keruspe** [comment]
@steve-chavez can you retry to see what you get with pull request #61 ?

**Comment by steve-chavez** [comment]
Sorry for the delay, @Keruspe when running my latest example(one without postgres) with your PR the logs(I'm pasting starting from the successful traces of declaring the queues, maybe it helps) now give:

```rust
TRACE:lapin_futures::transport: decoded frame: Method(1, Queue(DeclareOk(DeclareOk { queue: "test", message_count: 8, consumer_count: 0 })))
TRACE:tokio_io::framed_read: frame decoded from buffer    
TRACE:lapin_futures::transport: upstream poll gave frame: Method(1, Queue(DeclareOk(DeclareOk { queue: "test", message_count: 8, consumer_count: 0 })))
TRACE:lapin_async::connection: will handle frame: Method(1, Queue(DeclareOk(DeclareOk { queue: "test", message_count: 8, consumer_count: 0 })))
TRACE:lapin_futures::transport: stream poll                                         
TRACE:lapin_futures::transport: Sending heartbeat                                                                                                             
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Heartbeat(0)
TRACE:lapin_futures::transport: serialized frame: 8 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=8
TRACE:tokio_io::framed_write: framed transport flushed
TRACE:tokio_io::framed_read: attempting to decode a frame
DEBUG:tokio_core::reactor: consuming notification queue
DEBUG:tokio_core::reactor: scheduling direction for: 0
DEBUG:tokio_core::reactor: blocking
TRACE:lapin_futures::transport: upstream poll gave NotReady
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 4422 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 22964, tv_nsec: 420771050 }
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 8609 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 59, nanos: 912405748 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23024, tv_nsec: 333188016 }
TRACE:tokio_core::reactor: event Ready {Readable | Writable} Token(2)
DEBUG:tokio_core::reactor: notifying a task handle
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 59618 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 17041 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23024, tv_nsec: 333268021 }
TRACE:tokio_core::reactor: event Ready {Readable} Token(1)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 18967 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 2622 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23024, tv_nsec: 333315412 }
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 15564 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 60, nanos: 1857361 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23084, tv_nsec: 335190201 }
TRACE:tokio_core::reactor: event Ready {Readable | Writable} Token(2)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 46958 }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 59, nanos: 999018613 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23144, tv_nsec: 334258805 }
TRACE:tokio_core::reactor: event Ready {Readable | Writable | Error | Hup} Token(2)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 49610 }
My message
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 153, nanos: 402412249 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23297, tv_nsec: 736726928 }
TRACE:tokio_core::reactor: event Ready {Readable} Token(1)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 52384 }
Sending: My message
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Method(1, Basic(Publish(Publish { ticket: 0, exchange: "", routing_key: "test", mandatory: false, 
immediate: false })))
TRACE:lapin_futures::transport: serialized frame: 21 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=21
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 104, message: "Connection reset by peer" } }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 13648 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23297, tv_nsec: 738387875 }
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 14702 }
```

Now the error ```TRACE:tokio_core::reactor: event Ready {Readable | Writable | Error | Hup} Token(2)``` is logged as soon the 3 mins mark is hit, before sending the message("My message" in the logs). The logs then stop giving output. If I try to send messages after that, the "Broken pipe" message is shown:

```rust
Other Message
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 471, nanos: 130175359 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23768, tv_nsec: 868581839 }
TRACE:tokio_core::reactor: event Ready {Readable} Token(1)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 49234 }
Sending: Other Message
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Method(1, Basic(Publish(Publish { ticket: 0, exchange: "", routing_key: "test", mandatory: false, 
immediate: false })))
TRACE:lapin_futures::transport: serialized frame: 21 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=21
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 6718 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23768, tv_nsec: 869954033 }
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 9592 }
Another Message
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 5, nanos: 326648529 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23774, tv_nsec: 196614852 }
TRACE:tokio_core::reactor: event Ready {Readable} Token(1)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 51568 }
Sending: Another Message
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Method(1, Basic(Publish(Publish { ticket: 0, exchange: "", routing_key: "test", mandatory: false, 
immediate: false })))
TRACE:lapin_futures::transport: serialized frame: 21 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=21
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 32, message: "Broken pipe" } }
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 0, nanos: 6730 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 23774, tv_nsec: 198051755 }
DEBUG:tokio_core::reactor: loop process - 0 events, Duration { secs: 0, nanos: 10582 }
```

**Comment by Keruspe** [comment]
@steve-chavez Can you run that again with the logs I just added in the PR?

**Comment by Keruspe** [comment]
Ok, I found out what's wrong.
In your example, instead of
```
channel.basic_publish(blah);
Ok(())
```
You should be doing something like
```
channel.basic_publish(blah).map(|_| ())
```

**Comment by steve-chavez** [comment]
Upgraded to ```lapin-futures = "0.9.0"```and also changed the code to use ```map``` as suggested:

```rust
fn main() {
  env_logger::init().unwrap();
  let mut core = Core::new().unwrap();
  let handle = core.handle();
  let addr = "127.0.0.1:5672".parse().unwrap();

  core.run(
    TcpStream::connect(&addr, &handle).and_then(|stream| Client::connect(stream, &ConnectionOptions::default()))
    .and_then(|client| client.create_channel())
    .and_then(|channel|
      channel.queue_declare("test", &QueueDeclareOptions::default(), FieldTable::new())
      .and_then(move |_|
        stdin().for_each(move |string| {
          println!("Sending: {}", string);
          channel.basic_publish("", "test", string.as_bytes(), &BasicPublishOptions::default(), BasicProperties::default())
          .map(|_|())
        })
      )
    )
  ).unwrap();
}
```

But still getting the error after 3 mins with the difference that now the program exits instead of keep running, the log:

```rust
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 81570 }
My Message
DEBUG:tokio_core::reactor: loop poll - Duration { secs: 18, nanos: 619325554 }
DEBUG:tokio_core::reactor: loop time - Instant { tv_sec: 26031, tv_nsec: 356774800 }
TRACE:tokio_core::reactor: event Ready {Readable} Token(1)
DEBUG:tokio_core::reactor: loop process - 1 events, Duration { secs: 0, nanos: 36889 }
Sending: My Message
TRACE:lapin_futures::transport: send frames
TRACE:lapin_futures::transport: sink start send
TRACE:lapin_futures::transport: will encode and write frame: Method(1, Basic(Publish(Publish { ticket: 0, exchange: "", routing_key: "test", mandatory: false, 
immediate: false })))
TRACE:lapin_futures::transport: serialized frame: 21 bytes
TRACE:lapin_futures::transport: sink poll_complete
TRACE:tokio_io::framed_write: flushing framed transport
TRACE:tokio_io::framed_write: writing; remaining=21
ERROR:lapin_futures::transport: Failed to send frame: Error { repr: Os { code: 104, message: "Connection reset by peer" } }
TRACE:lapin_futures::channel: Failed to handle frames: Error { repr: Os { code: 104, message: "Connection reset by peer" } }
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: ConnectionAborted, error: StringError("Failed to ha
ndle frames: Error { repr: Os { code: 104, message: \"Connection reset by peer\" } }") }) }', src/libcore/result.rs:859
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by Keruspe** [comment]
Yep, this is the expected behaviour. If your tcp connection is failing, the error is now successfully forwarded and thus your program exits.

**Comment by steve-chavez** [comment]
But shouldn't the tcp connection be kept alive by the heartbeat? In contrast of my publishing example, running a basic consume example with 0.9.0 doesn't yield this error after the mentioned 3 mins, it just keeps running.

**Comment by Keruspe** [comment]
That indeed, I think I misunderstood what your actual problem was.
You problem is that the heartbeat is no triggerred when you program is idle, right?

**Comment by steve-chavez** [comment]
Yes, that's my problem as per my initial comment after 3 mins of being idle the connection is not kept alive by the heartbeat.

**Comment by Keruspe** [comment]
Yup, I just thought that you were saying that the error appeared after 3 minutes of the program only sending heartbeats, I didn't get that no heartbeat were sent.
I can reproduce now. (I used to just restart the rabbitmq-server to be able to fake a network issue)

**Comment by Keruspe** [comment]
I need to find a way to do something like that cleanly in lapin itself but in the meantime this workaround should "fix" you problem: https://github.com/Keruspe/lapin/commit/41682f3c47ec3c9aaa00d6baed8bfd9438c9358d

**Comment by Keruspe** [comment]
This should be fixed in #63

**Comment by steve-chavez** [comment]
Thanks a lot @Keruspe, I've tried your master branch with the heartbeat background thread and now the tcp connection is kept alive after the 3 mins of being idle.
